### ISSUE TITLE
Spark 3.0.1 No Encoder found for com.google.protobuf.ByteString

### ISSUE THREAD

**Issue by guoch** [scalapb/sparksql-scalapb#154]
Hi Team,

I am getting the  No Encoder found for com.google.protobuf.ByteString in the sparksql-scalapb 0.11.0-RC1 with Spark3.
I used this example https://github.com/thesamet/sparksql-scalapb-test with newer package version, and 
https://scalapb.github.io/docs/sparksql#from-binary-to-protos-and-back, shaded the copy of the Protocol Buffer runtime and scala-collection-compat as well, but it didn't help
Sharing my code in the zip  [protoscalapb.zip](https://github.com/scalapb/sparksql-scalapb/files/5780453/protoscalapb.zip)

Environment:  
openjdk version "1.8.0_275" 
scala 2.12.10,
Spark 3.0.1
"sparksql-scalapb" % "0.11.0-RC1"

Reproduce Steps:
===================
Download the Open Source Spark 3.0.1,
Download my attachment, and  sbt assembly to generate the jar, attach to spark-shell
```
wget https://downloads.apache.org/spark/spark-3.0.1/spark-3.0.1-bin-hadoop2.7.tgz
tar -zxvf spark-3.0.1-bin-hadoop2.7.tgz
spark-3.0.1-bin-hadoop2.7/bin/spark-shell --jars ~/protoscalapb-assembly-0.1.jar
```

Run the below code in the spark-shell
```
import org.apache.spark.sql.{Dataset, DataFrame, functions => F}
import spark.implicits.StringToColumn
import scalapb.spark.ProtoSQL
import scalapb.spark.Implicits._
import protobuf.test.Person

val testData = Seq(
   Person(name="chg",id=32),
   Person(name="Andrew", id=29)
)
val binaryDS: Dataset[Array[Byte]] = spark.createDataset(testData.map(_.toByteArray))
val protosDS: Dataset[Person] = binaryDS.map(Person.parseFrom(_))
```

```
java.lang.UnsupportedOperationException: No Encoder found for com.google.protobuf.ByteString
- array element class: "com.google.protobuf.ByteString"
- field (class: "scala.collection.Seq", name: "lengthDelimited")
- map value class: "scalapb.UnknownFieldSet.Field"
- field (class: "scala.collection.immutable.Map", name: "fields")
- field (class: "scalapb.UnknownFieldSet", name: "unknownFields")
- root class: "protobuf.test.Person"
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$serializerFor$1(ScalaReflection.scala:567)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:882)
................
................
  at org.apache.spark.sql.Encoders$.product(Encoders.scala:285)
  at org.apache.spark.sql.LowPrioritySQLImplicits.newProductEncoder(SQLImplicits.scala:251)
  at org.apache.spark.sql.LowPrioritySQLImplicits.newProductEncoder$(SQLImplicits.scala:251)
  at org.apache.spark.sql.SQLImplicits.newProductEncoder(SQLImplicits.scala:32)
  ... 49 elided

```
Could you please help take a look?  Thank you in advance.

Regards

**Comment by thesamet** [comment]
Hi @guoch , can you try using ScalaPB 0.10.x. Version 0.11.x is not tested with SparkSQL-scalapb. Please see the table in this doc for supported combinations of versions: https://scalapb.github.io/docs/sparksql

Another possibility for the problem you are seeing is that spark-shell is bringing in Spark's default implicits. If the problem persists after downgrading ScalaPB to 0.10.x, try deploying the same code locally instead of typing it in spark-shell.

**Comment by guoch** [comment]
Hi @thesamet  Thank you for looking into this.  I also tried the scala PB "sparksql-scalapb" % "0.10.4", it doesn't work as well. Failed with the same error trace. Below is my build.sbt

==========================
```
name := "protoscalapb"

version := "0.2"

scalaVersion := "2.12.10"

libraryDependencies ++= Seq(
  "com.thesamet.scalapb" %% "scalapb-runtime" % scalapb.compiler.Version.scalapbVersion % "protobuf",
  "com.thesamet.scalapb" %% "sparksql-scalapb" % "0.10.4"
)


assemblyShadeRules in assembly := Seq(
  ShadeRule.rename("com.google.protobuf.**" -> "shadeproto.@1").inAll,
  ShadeRule.rename("scala.collection.compat.**" -> "shadecompat.@1").inAll
)

PB.targets in Compile := Seq(
  scalapb.gen() -> (sourceManaged in Compile).value / "scalapb"
)


```

**Comment by guoch** [comment]
In addition, I previously tested it with a jar installed to a Databricks Runtime 7.4 (Spark 3.0.1), that Jar installed to all the executors as well.  It fails, then I tried to test in the Open Source Spark 3.0.1.

In the error trace, I see it is looking for class: "com.google.protobuf.ByteString",   My understanding is that, I already shade rename  , should the catalyst looks for the  shadeproto.ByteString instead?

**Comment by guoch** [comment]
Hi @thesamet 

I verify that the cause might be the "spark-shell is bringing in Spark's default implicits" as you mentioned.
Verified that the spark-submit is the only working way 
`spark-3.0.1-bin-hadoop2.7/bin/spark-submit --class myexample.RunDemo protoscalapb-assembly-0.3.jar`

The situation seems to be related with the SparkSession.builder().getOrCreate().  For spark-shell and Databricks platform,  they use the shared SparkContext API to get the SparkContext, this get the existing spark context already instead of creating a new sparkcontext, and then the "com.google.protobuf.ByteString" could not be found.  

In this way, if we try to deploy it in Databricks or similar product, it won't be possible to make it run `val protosDS: Dataset[Person] = binaryDS.map(Person.parseFrom(_)) `  unless using spark-submit.

Not sure if any feature to make it work in an existing SparkContext in the future?

**Comment by thesamet** [comment]
Hi @guoch , the issue is not related to where the SparkContext is created (or whether it is shared) , it is related to which `implicit` definitions are imported in the scope where your code is running. It looks like spark-shell has it [hard-coded]( https://github.com/apache/spark/blob/1495ad8c46197916527236331b57dce93aa3b8ec/repl/src/main/scala-2.12/org/apache/spark/repl/SparkILoop.scala#L74). I am not sure about databricks, but it's worth asking or making a feature request for them to disable those forced imports.

**Comment by guoch** [comment]
Sure, thank you for your clarification. I will also check the Databricks/Spark developers to research further.

**Comment by guoch** [comment]
I talked to one of the guy from Databricks product group, they confirmed that they won't disable those forced imports in their platform.  Looks like the databricks notebook is doing similar thing to the spark-shell.

Databricks also has spark-submit, I tested that work as well.

So overall, the cause is related with implicit definitions scope,   It can't work in Databricks platform or spark-shell.  Can only work if we package into a jar and use  spark-submit to do that.    Since this is by design, I think the issue can be closed.  Thank you for your help.
